### PR TITLE
Fix access to unitialized memory in RNG ops

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Random operators used vector of reinterpreted bytes as a target of assignment when creating distribution objects.
The random choice uses non-trivial `std::discrete_distribution` carrying some state. Assigning to that memory caused a leak.
Switch to a proper `vector<distribution>`, so the copy/move assignments are not UB as the location that we assign to is now
properly constructed ahead of time.

### Affected modules and functionalities:
RNG operators

### Tests:

- [x] Existing tests apply
The relevant fn.random.choice tests no longer report leak under ASAN.
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/operators/random/rng_base_cpu.h [hunk 1]
```diff
@@ -125,10 +123,8 @@ void RNGBase<Backend, Impl, IsNoiseGen>::RunImplTyped(Workspace &ws, CPUBackend)
 
   // TODO(janton): set layout explicitly from the user for RNG
 
-  auto &dists_cpu = backend_data_.dists_cpu_;
-  dists_cpu.resize(sizeof(Dist) * nsamples);  // memory was already reserved in the constructor
-  Dist* dists = reinterpret_cast<Dist*>(dists_cpu.data());
-  bool use_default_dist = !This().template SetupDists<T>(dists, ws, nsamples);
+  std::vector<Dist> dists(nsamples);
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/operators/random/rng_base_cpu.h:126]: This is just a batch of the default constructed objects, which should be cheap, and we still reassign them in every iteration in most cases. Based on this I thought it should be ok.

**Review by JanuszL — APPROVED** [review]

**Review by stiepan — APPROVED** [review]

File: dali/operators/random/rng_base_cpu.h (hunk 1 above)
Comment (mzient) [inline dali/operators/random/rng_base_cpu.h:126]: This is captured by value.

**Review by mzient — CHANGES_REQUESTED** [review]
Please capture the distribution vector by reference or return to the use of a pointer.

**Comment by klecki** [comment]
Same fix with less allocations reimplemented in #5395

**Pull request closed without merging**